File: python/algo/202505/algo1977.py
from functools import cache

# 动态规划 - 11.1 前缀和优化 DP
class Solution:
    def numberOfCombinations(self, num: str) -> int:
        # 自行解答 - 记忆化搜索
        mod = 10 ** 9 + 7
        n = len(num)

        @cache
        def dfs(i:int, pre:int) -> int:
            if i == n: return 1
            if num[i] == '0': return 0

            res = 0
            for j in range(max(i-pre,1), n-i+1):
                if (j == i - pre and num[i:i+j]<num[pre:i]):continue
                res += dfs(i+j,i)

            return res % mod
        return dfs(0, 0)


    def numberOfCombinations2(self, num: str) -> int:
        # 自行解答 - 递归 -> 递推
        if num[0] == '0': return 0
        mod = 10 ** 9 + 7
        n = len(num)

        dp = [[0] * (n) for _ in range(n+1)]
        for i in range(n): dp[n][i] = 1

        for i in range(n-1, -1, -1):
            if num[i] == '0':
                for j in range(n): dp[i][j] = 0
                continue
            for j in range(max(0,2*i-n), i+1):
                for k in range(max(i-j,1), n-i+1):
                    if (k == i-j and num[i:i+k]<num[j:i]):continue
                    dp[i][j] += dp[i+k][i]

        return dp[0][0] % mod

    def numberOfCombinations3(self, num: str) -> int:
        # 自行解答 - 前缀和优化
        if num[0] == '0': return 0
        mod = 10 ** 9 + 7
        n = len(num)

        dp = [[0] * (n) for _ in range(n+1)]
        for i in range(n): dp[n][i] = 1

        for i in range(n-1, -1, -1):
            if num[i] == '0':
                for j in range(n): dp[i][j] = 0
                continue
            sum = 0
            
            for k in range(i+1, n-i+1): sum += dp[i+k][i]
            for j in range(max(0,2*i-n), i+1):
                is_large = num[i:i+i-j]>=num[j:i]
                if (is_large):
                    sum += dp[i+i-j][i]
                dp[i][j] += sum
                if (not is_large):
                    sum += dp[i+i-j][i]

        return dp[0][0] % mod

if __name__ == "__main__":
    sol = Solution()
    print(sol.numberOfCombinations("327"))

File: cplus/leetcode/202505/algo1977.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 - 11.1 前缀和优化 DP 难度分：2817
class Solution {
public:
    int numberOfCombinations(string num) {
        // 自行解答 记忆化搜索 -> 递推 见 python
        // 直接从 python 转换过来，memory overtime
        if (num[0] == '0') return 0;

        const int mod = 1e9 + 7, n = num.size();
        vector<vector<int>> dp(n+1, vector<int>(n));
        
        for (int j=0; j<n; ++j) dp[n][j] = 1;
        for (int i=n-1; i>=0; --i) {
            if (num[i] == '0') {
                for (int j=0; j<n; ++j) dp[i][j] = 0;
                continue;
            }
            int sum = 0;
            for (int k=i+1; k<=n-i; ++k) sum += dp[i+k][i];

            for (int j=max(0, 2*i-n); j<=i; ++j) {
                if (num.substr(i, i-j) >= num.substr(j,i-j))
                    sum = ((long)dp[i+i-j][i] + sum) % mod;
                dp[i][j] = ((long)dp[i][j] + sum) % mod;
                if (num.substr(i, i-j) < num.substr(j,i-j))
                    sum = ((long)dp[i+i-j][i] + sum) % mod;

            }
        }

        return dp[0][0];
    }

    int numberOfCombinations2(string num) {
        if (num[0] == '0') return 0;

        const int mod = 1e9 + 7, n = num.size();
        
        // c++ 需要计算预处理最长公共前缀
        // 最长公共前缀 - 朴素算法
        // lcp[i][j] 表示以 i, j 开始的子串的最长公共前缀的长度
        // lcp[i][j] = lcp[i+1][j+1] + 1, num[i] == num[j]
        //             0                , num[i] != num[j]
        vector<vector<int>> lcp(n, vector<int>(n));
        for (int i=n-1; i>=0; --i) {
            // 以通过画图理解避免 j+1 越界，!!很棒的写法!!
            lcp[i][n-1] = (num[i] == num[n-1]); 
            for (int j=i+1; j<n-1; ++j) {
                lcp[i][j] = (num[i]==num[j])?lcp[i+1][j+1]+1:0;
            }
        }
        // 判断逻辑是 r >= l:true, r<l false
        auto cmp = [&](int l, int r, int len) -> bool {
            if (lcp[l][r] >= len) return true;                      // l == r
            if (num[l+lcp[l][r]] > num[r+lcp[l][r]]) return false;  // l > r
            else return true;                                       // l < r
        };


        vector<vector<int>> dp(n+1, vector<int>(n));
        
        for (int j=0; j<n; ++j) dp[n][j] = 1;
        for (int i=n-1; i>=0; --i) {
            if (num[i] == '0') {
                for (int j=0; j<n; ++j) dp[i][j] = 0;
                continue;
            }
            int sum = 0;
            for (int k=i+1; k<=n-i; ++k) sum = ((long)dp[i+k][i] + sum) % mod;

            for (int j=max(0, 2*i-n); j<=i; ++j) {
                // bool is_large = (num.substr(i, i-j) >= num.substr(j,i-j))
                bool is_large = cmp(j, i, i-j);
                if (is_large)
                    sum = ((long)dp[i+i-j][i] + sum) % mod;
                dp[i][j] = ((long)dp[i][j] + sum) % mod;
                if (!is_large)
                    sum = ((long)dp[i+i-j][i] + sum) % mod;

            }
        }

        return dp[0][0];
    }
};